Update coreclr test run scripts

BruceForstall opened the pull request:
1. Add a run.py `-logs_dir` argument (run.cmd: `logsdir`; run.sh: `-logsDir`) to allow specifying a particular logs directory name. By default, use "logs", as before. This can be used to help preserve per-test-run output (note, however, that currently many per-test-run assets litter the tests artifacts directories). This can be useful if you have multiple test run results logs directory and you want to use `--analyze_results_only` (run.cmd: `printlastresultsonly`) to analyze a run that is not the default logs directory.
2. Remove the unnecessary run.py `--skip_test_run` argument. It was added to use the logic to precompile the libraries, but that logic is now gone.
3. Move generated repro files for failed tests to the "log\repro" directory, and name them with a `repro_` filename prefix.
4. Fix repro file generation. Currently, it gives up on creating repro files for merged tests (it doesn't know how to generate a repro file). Repro files now work for `BuildAsStandalone=true` test builds.
5. Print out the path of all repro files that are created.
6. Add a `setlocal` to the generated Windows repro files.
7. Misc. minor improvements.

Contributes to #95032

BruceForstall committed:
Update coreclr test run scripts

1. When running tests, create a new unique 'logs' directory for every run,
to avoid overwriting the results of previous runs. First, use "artifacts\log".
Then, use "log.1", "log.2", etc.
2. Add a run.py `-logs_dir` argument (run.cmd: `logsdir`; run.sh: `-logsDir`) to
allow specifying a particular logs directory name. This can be useful if you have
multiple test run results logs directory and you want to use `--analyze_results_only`
(run.cmd: `printlastresultsonly`) to analyze a run that is not the last logs directory.
3. Remove the unnecessary run.py `--skip_test_run` argument. It was added to use the logic
to precompile the libraries, but that logic is now gone.
4. Move generated repro files for failed tests to the "log\repro" directory, and name them
with a `repro_` filename prefix.
5. Print out the path of all repro files that are created.
6. Add a `setlocal` to the generated Windows repro files.
7. Misc. minor improvements.

Contributes to #95032

BruceForstall committed:
Try to fix Linux repros

BruceForstall commented:
@trylek @jkoritzinsky @markples I'm trying to fix run.py, especially the ability to automatically create repro scripts for failing tests. It works now for BuildAsStandalone=true, but not for merged tests.

In a merged test, we end up with an xml file like JIT.SIMD.JIT.SIMD.testRun.xml in the log directory (it looks like the testRun.xml file is copied to the log directory, replacing path separators with periods. Aside: couldn't we copy it preserving the same relative path and not munge the filename?), with contents like:
```
<assemblies>

<assembly
    name="JIT.SIMD"
    test-framework="XUnitWrapperGenerator-generated-runner"
    run-date="2023-11-26"
    run-time="22:40:26"
    time="14.6475557"
    total="116"
    passed="116"
    failed="0"
    skipped="0"
    errors="0">

<collection
    name="Collection"
    time="14.6475557"
    total="116"
    passed="116"
    failed="0"
    skipped="0"
    errors="0"
>
<test name="JIT\SIMD\AbsGeneric_r\AbsGeneric_r.dll" type="VectorMathTests.Program" method="TestEntryPoint" time="0.123840" result="Pass"></test>

<test name="JIT\SIMD\AbsGeneric_ro\AbsGeneric_ro.dll" type="VectorMathTests.Program" method="TestEntryPoint" time="0.002934" result="Pass"></test>

<test name="JIT\SIMD\AbsSqrt_r\AbsSqrt_r.dll" type="VectorMathTests.Program" method="TestEntryPoint" time="0.004617" result="Pass"></test>

<test name="JIT\SIMD\AbsSqrt_ro\AbsSqrt_ro.dll" type="VectorMathTests.Program" method="TestEntryPoint" time="0.000910" result="Pass"></test>

<test name="JIT\SIMD\AddingSequence_r\AddingSequence_r.dll" type="VectorMathTests.Program" method="TestEntryPoint" time="0.003844" result="Pass"></test>

<test name="JIT\SIMD\AddingSequence_ro\AddingSequence_ro.dll" type="VectorMathTests.Program" method="TestEntryPoint" time="0.002271" result="Pass"></test>

<test name="_BitwiseOperations_r::VectorMathTests.Program.TestDouble()" type="VectorMathTests.Program" method="TestDouble" time="0.005970" result="Pass"></test>

<test name="_BitwiseOperations_r::VectorMathTests.Program.TestBool()" type="VectorMathTests.Program" method="TestBool" time="0.008456" result="Pass"></test>

<test name="_BitwiseOperations_ro::VectorMathTests.Program.TestDouble()" type="VectorMathTests.Program" method="TestDouble" time="0.003036" result="Pass"></test>

<test name="_BitwiseOperations_ro::VectorMathTests.Program.TestBool()" type="VectorMathTests.Program" method="TestBool" time="0.002500" result="Pass"></test>
...
```

How do we convert this information into a script to run the test, for each test that has `result="Fail"`? It looks like the `<assembly>` "name" is the name of the wrapper assembly (minus the ".dll" extension). So assuming these are all unique, we can just search the tree for it. Then invoke `corerun JIT.SIMD.dll XXX` where `XXX` is the `<test name="XXX"...`? I tried that and it seems to work, but then again, `corerun JIT.SIMD.dll foobar` doesn't fail, and in fact, returns a 100 return code and produces no output or error message, so I'm not confident.

janvorli commented:
Adding the index to the log directory means that I cannot use the same shell path to search in the generated logs. I always rely on shell history to run stuff from the command line, so I'd have to modify such a search command unless I deleted the log folder before every run. It seems that we could change the naming schema as follows. If there was a log folder already, then rename it to log.{datetime} where the datetime would be the file creation or modification time. Then just create the log folder without any suffix and use it for the new run.

trylek commented:
@BruceForstall - if you're suggesting basically creating a subtree with the xml results under ```artifacts\log```, that should work even though we'd need to modify the run.py script to collect the xml files across the entire subtree, probably not super hard. Alternatively we could probably somehow encode the relative path to the merged wrapper in some new attribute of the ```assembly``` tag. It might also be possible to generate the repro scripts immediately when building the tests, emitting a single-line file should take a negligible amount of time in comparison with all the msbuild and other machinery that happens under the bonnet. I'm curious whether @jkoritzinsky or @markples might have additional ideas.

For @janvorli's comment, I like the idea of renaming the log to include a timestamp when rerunning the tests; having said that, I suggested something similar in

https://github.com/dotnet/runtime/issues/90682

and several people including JanK expressed concerns towards such a scheme so these would need reaching an agreement on if we decide to proceed in this direction. Other than that I'm not sure what exactly you meant by ```index```, if I understand Bruce's suggestion correctly he basically proposed putting the xml results into subfolders so that their paths relative to ```artifacts\log``` would let him easily identify the relative path of the corresponding merged wrapper under ```artifacts\tests\coreclr\os.arch.config```; please clarify or correct me in case that was just my misunderstanding.

trylek commented:
@BruceForstall - IIRC it should suffice to run the merged wrapper passing in the test in question as the first command-line argument, @jkoritzinsky should confirm. According to my recollection there should be a minimalistic set of ```if``` statements generated into the merged wrapper to support narrowing its execution to a single test, when Jeremy wrote this logic he was aware of the general ask for minimum overhead originally raised by the JIT team. If that doesn't work or doesn't meet expectations, it should be relatively simple to update the merged wrapper generator to take this into consideration.

I also think you're right that the merged wrapper always returns 100 even when some tests failed, otherwise Helix nukes the work item without uploading any results AFAIK, that will likely need fixing for this execution mode if we want the merged wrapper to return the actual test exit code in this case - in a way, it seems to me that passing in the test name should internally turn the merged wrapper functionality into that of the "simple wrapper" used for out-of-process tests. According to my knowledge Jeremy is back in office now, once we decide on what needs to be done I'll be happy to lend a hand to get this fixed in the next couple of days.

markples commented:
See #93452 about specifying tests.

Also, just in case, there was a bug in the `.cmd` scripts where passing a test name to the script didn't work.  This was fixed in July, but you (if like me) might have a memory of not being able to get this to work at all.

I believe that having "passing the test name" change the exit behavior may be difficult to reconcile because the argument is a general filter argument.  In theory this could end up being used for various tasks (e.g., use filters instead of compiling directories into multiple merged groups?).  But I assume that the merged test script is called from some other script that could behave differently (or we could add a layer if the "next layer up" is Helix and difficult to change)?

BruceForstall committed:
Remove automatic creation of new 'log' directories

Also, fix repro creation for non-BuildAsStandalone by
bailing; we don't know how to create those repros.

Fix repro creation for BuildAsStandalone.

Improve some output messages and code comments.

BruceForstall commented:
@trylek @janvorli I removed the code that created new 'log' directories, so now it uses the same "artifacts/log" directory as before. Mostly, the change now is to fix repro file generation, plus clean up some code (add comments, e.g.). I left the "-logs_dir" arguments since that could potentially still be useful either now or in the future.
